### PR TITLE
feat: replace frequency-based validation with configurable event-based system

### PR DESCRIPTION
**Pull request by RobotSail**
## Summary

- Adds `ValidationScheduler` class mirroring the existing `Checkpointer`'s event-based design pattern
- Validation can now be triggered by four configurable event types:
  - **step**: validate every N training steps (existing `--validation-frequency` behavior, preserved for backward compatibility)
  - **epoch**: validate at end of each epoch (`--validate-at-epoch`)
  - **samples**: validate every N accumulated samples (`--min-samples-per-validation`)
  - **final**: validate at end of training (`--validate-at-final`)
- The existing `--validation-frequency` remains as the step-based trigger — no breaking changes
- At least one validation trigger must be configured when validation data is provided (previously only `validation_frequency` was accepted)
- `ValidationScheduler` state is saved/restored in full-state checkpoints for correct resume behavior
- Epoch-based and final validation also trigger `best_val_loss` checkpoint saves when applicable

Closes #MT-80

## Test plan

- [x] 29 new unit tests for `ValidationScheduler` covering all 4 event types, state tracking, edge cases, and multi-trigger scenarios
- [x] New API test for event-based validation parameter forwarding (`test_run_training_event_based_validation`)
- [x] Full non-GPU test suite passes: 367 passed, 1 skipped (2 pre-existing OSFT failures unrelated to this change)
- [x] Ruff lint + format checks pass
- [ ] GPU validation with `model_validation.py` pending (to be run by validator agent)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added event-driven validation scheduling (by steps, epochs, processed samples, or final completion) with configurable triggers.
  - Expanded training/CLI configuration for epoch-based, minimum-sample-count, and final validation options.
  - Validation progress/state is saved and restored in full-state checkpoints, preserving resume behavior.
  - Multiple applicable triggers are coalesced so validation runs at most once per step.

- **Bug Fixes**
  - Now raises an error when validation data is provided without any validation trigger configured.
  - Enforces `min_samples_per_validation` as an integer ≥ 1.

- **Documentation**
  - Updated validation docs and CLI parameter descriptions to reflect event-based validation and checkpoint behavior.

### CODE REVIEW

File: src/mini_trainer/train.py [hunk 1]
```diff
@@ -662,6 +663,84 @@ def record_save(
...
+        self.last_sample_based_validation_samples = 0
+
+    @property
+    def is_configured(self) -> bool:
+        return (
+            self.validation_frequency is not None
+            or self.validate_at_epoch
+            or self.min_samples_per_validation is not None
+            or self.validate_at_final
+        )
```
[the remark this answers is not in the document]
Comment (RobotSail) [reply src/mini_trainer/train.py:707]: Fixed in commit 04a84fb: `ValidationScheduler.__init__` now normalizes `validation_frequency < 1` to `None`, and `is_configured` checks `> 0`. Both are now consistent with `should_validate`.

**Comment by RobotSail** [comment]
Addressed in commit 04a84fb: `ValidationScheduler.__init__` now normalizes `validation_frequency < 1` to `None`, and `is_configured` checks `> 0`. Both are now consistent with `should_validate`.

File: src/mini_trainer/api_train.py [hunk 2]
```diff
@@ -156,8 +156,29 @@ def run_training(torch_args: TorchrunArgs, train_args: TrainingArgs) -> None:
...
+        )
+        if not has_trigger:
+            raise ValueError(
+                "Validation data is provided but no validation trigger is configured. "
+                "Set at least one of: validation_frequency, validate_at_epoch, "
+                "min_samples_per_validation, or validate_at_final."
+            )
+        if train_args.min_samples_per_validation is not None and train_args.min_samples_per_validation < 1:
+            raise ValueError("min_samples_per_validation must be a positive integer when set")
+
```
[the remark this answers is not in the document]
Comment (RobotSail) [reply src/mini_trainer/api_train.py:173]: Fixed in commit 7e678ac: bounds validation now runs before the trigger check in both `api_train.py` and `train.py`. Tests updated to verify `min_samples_per_validation=0` raises the actionable error even as the sole trigger param.

File: Readme.md [hunk 3]
```diff
@@ -124,6 +125,11 @@ torchrun --nnodes=1 --nproc-per-node=8 -m mini_trainer.train \
 - `--osft` - Enable Orthogonal Subspace Fine-Tuning mode
 - `--osft-unfreeze-rank-ratio` - Ratio of model parameters to train with OSFT (0.0-1.0)
 - `--block-size` - Enables pretraining mode with the given block length
+- `--validation-split` - Fraction of training data to hold out for validation (0.0-1.0)
```
[the remark this answers is not in the document]
Comment (RobotSail) [reply Readme.md]: Fixed in commit a34123d: changed from `(0.0-1.0)` to `(greater than 0, less than 1)` to make the exclusive bounds explicit.

File: Readme.md [hunk 4]
```diff
@@ -124,6 +125,11 @@ torchrun --nnodes=1 --nproc-per-node=8 -m mini_trainer.train \
...
+- `--validation-data-path` - Path to a separate validation dataset (mutually exclusive with `--validation-split`)
+- `--validation-frequency` - Run validation every N steps
+- `--validate-at-epoch` - Run validation at the end of each epoch
+- `--validate-at-final` - Run validation at the end of training
```
[the remark this answers is not in the document]
Comment (RobotSail) [reply Readme.md]: Fixed in commit a34123d: added `--min-samples-per-validation` to the Key Parameters section.
